File: leetcode/medium/RestoreIpAddresses.py
from typing import List

class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        """
        Restores all possible valid IP address combinations from a given string `s`.
        A valid IP address consists of exactly four integers, each integer ranging from 0 to 255,
        separated by single dots. Each integer cannot have leading zeros unless it is the number 0 itself.

        Args:
            s: The input string containing only digits.

        Returns:
            A list of all possible valid IP address strings.
        """
        
        n = len(s)
        # A valid IP address has a minimum length of 4 (e.g., "0.0.0.0")
        # and a maximum length of 12 (e.g., "255.255.255.255").
        if not (4 <= n <= 12):
            return []
        
        result = [] # This list will store all valid IP addresses found.

        # The backtracking function to explore combinations.
        # `start_index`: The starting index in the string `s` for the current segment.
        # `segments`: A list storing the current IP address segments.
        def backtrack(start_index: int, segments: List[str]):
            # Base case 1: If we have formed 4 segments.
            if len(segments) == 4:
                # If we have used all characters in `s`, then it's a valid IP address.
                if start_index == n:
                    result.append(".".join(segments))
                return # Either valid and added, or invalid (not all chars used).

            # Base case 2: If we have used all characters but haven't formed 4 segments.
            if start_index == n:
                return

            # Try to form the next segment (1, 2, or 3 digits long).
            # A segment can be 1 digit long.
            if start_index < n:
                segment = s[start_index : start_index + 1]
                segments.append(segment)
                backtrack(start_index + 1, segments)
                segments.pop() # Backtrack

            # A segment can be 2 digits long.
            # It cannot have a leading zero (e.g., "01" is invalid).
            if start_index + 1 < n and s[start_index] != '0':
                segment = s[start_index : start_index + 2]
                segments.append(segment)
                backtrack(start_index + 2, segments)
                segments.pop() # Backtrack

            # A segment can be 3 digits long.
            # It cannot have a leading zero and its value must be between 0 and 255.
            if start_index + 2 < n and s[start_index] != '0':
                segment = s[start_index : start_index + 3]
                if 0 <= int(segment) <= 255:
                    segments.append(segment)
                    backtrack(start_index + 3, segments)
                    segments.pop() # Backtrack

        # Start the backtracking process from the beginning of the string with an empty list of segments.
        backtrack(0, [])
        return result            

File: leetcode/medium/RestoreIpAddresses.java
import java.util.ArrayList;
import java.util.List;

/**
 * Problem: Restore IP Addresses
 * Given a string containing only digits, restore it by returning all possible valid IP address combinations.
 * A valid IP address consists of exactly four octets (0-255) separated by dots, with no leading zeros 
 * (unless the segment is exactly "0").
 */
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> result = new ArrayList<>();
        int n = s.length();
        
        // An IPv4 address has between 4 and 12 digits.
        if (n < 4 || n > 12) return result;
        
        backtrack(s, 0, 0, new StringBuilder(), result);
        return result;
    }

    /**
     * @param s          The original input string.
     * @param start      The current index in the input string.
     * @param octetCount The number of octets already placed (0 to 4).
     * @param current    The current IP string being built.
     * @param result     The list to store valid IP addresses.
     */
    private void backtrack(String s, int start, int octetCount, StringBuilder current, List<String> result) {
        int remainingChars = s.length() - start;
        int remainingOctets = 4 - octetCount;

        // Pruning: Check if the remaining characters can fit into the remaining octets.
        // Each octet needs 1-3 characters.
        if (remainingChars < remainingOctets || remainingChars > remainingOctets * 3) {
            return;
        }

        // Base case: If we have 4 octets and have used all characters.
        if (octetCount == 4) {
            if (start == s.length()) {
                // Remove the trailing dot before adding to results.
                result.add(current.substring(0, current.length() - 1));
            }
            return;
        }

        int prevLen = current.length();

        // Try segments of length 1, 2, and 3.
        for (int len = 1; len <= 3 && start + len <= s.length(); len++) {
            String segment = s.substring(start, start + len);
            
            if (isValid(segment)) {
                current.append(segment).append('.');
                backtrack(s, start + len, octetCount + 1, current, result);
                current.setLength(prevLen); // Backtrack: reset StringBuilder length
            }
        }
    }

    /**
     * Helper to validate an IP segment (0-255).
     * Rules: No leading zeros unless segment is "0", and value <= 255.
     */
    private boolean isValid(String segment) {
        if (segment.length() > 1 && segment.startsWith("0")) return false;
        int val = Integer.parseInt(segment);
        return val <= 255;
    }
}